### PR TITLE
RandomX: update to v1.1.4

### PR DESCRIPTION
**Pull request by tevador**
* Faster cache initialization with SSSE3/AVX2
* Automatic detection of CPU capabilities in RandomX
* Fixed a possible out-of-bounds access in superscalar program generator
* Use MONERO_RANDOMX_UMASK to manually disable RandomX flags in monerod

### CODE REVIEW

File: src/crypto/rx-slow-hash.c [hunk 1]
```diff
@@ -282,14 +318,9 @@ void rx_slow_hash(const uint64_t mainheight, const uint64_t seedheight, const ch
     memcpy(rx_sp->rs_hash, seedhash, HASH_SIZE);
   }
   if (rx_vm == NULL) {
-    randomx_flags flags = RANDOMX_FLAG_DEFAULT;
-    if (use_rx_jit()) {
-      flags |= RANDOMX_FLAG_JIT;
-      if (!miners)
-          flags |= RANDOMX_FLAG_SECURE;
+    if ((flags & RANDOMX_FLAG_JIT) && !miners) {
+        flags |= RANDOMX_FLAG_SECURE;
```
Comment (moneromooo-monero) [inline src/crypto/rx-slow-hash.c]: Should this not be &'d against ~disabled flags ?

Comment (tevador) [reply src/crypto/rx-slow-hash.c]: Good point. Updated.

File: src/crypto/rx-slow-hash.c [hunk 2]
```diff
@@ -154,6 +155,43 @@ static inline int use_rx_jit(void)
...
+    return flags;
+  }
+
+  const char *env = getenv("MONERO_RANDOMX_UMASK");
+  if (!env) {
+    flags = 0;
+  }
+  else {
+    char* endptr;
+    long int value = strtol(env, &endptr, 10);
```
Comment (hyc) [inline src/crypto/rx-slow-hash.c]: I would use a base of 0 here, and allow the value to use other than base 10. Personally I would always use 0x....

File: src/crypto/rx-slow-hash.c [hunk 3]
```diff
@@ -236,7 +274,7 @@ void rx_slow_hash(const uint64_t mainheight, const uint64_t seedheight, const ch
   char *hash, int miners, int is_alt) {
   uint64_t s_height = rx_seedheight(mainheight);
   int toggle = (s_height & SEEDHASH_EPOCH_BLOCKS) != 0;
-  randomx_flags flags = RANDOMX_FLAG_DEFAULT;
+  randomx_flags flags = enabled_flags() & ~disabled_flags();
```
Comment (hyc) [inline src/crypto/rx-slow-hash.c]: We can delete cpuid() and check_aes_hw() now, nothing calls them. (Lines 103-131.)

File: src/crypto/rx-slow-hash.c (hunk 2 above)
Comment (tevador) [reply src/crypto/rx-slow-hash.c]: OK, changed the base to `0`.

File: src/crypto/rx-slow-hash.c (hunk 3 above)
Comment (tevador) [reply src/crypto/rx-slow-hash.c]: I have removed the unused functions.

**Comment by tevador** [comment]
BTW I added the option to disable mining with the dataset (lines 244-246). It has the same effect as if dataset allocation failed. May be useful if someone can't afford the additional 2 GB for mining.

``` 
MONERO_RANDOMX_UMASK=4 ./monerod
```

**Comment by hyc** [comment]
Do we just tell people to lookup the flags in randomx.h, or can you provide other documentation for that?

**Comment by tevador** [comment]
@hyc AFAIK none of the ENV flags were documented before. At least I can't find anything in the monero docs.

|flag|effect|
|----|------|
|2|force software AES|
|4|force mining in low memory mode|
|8|force the interpreter mode|
|16|disable secure JIT|
|32|disable SSSE3 Argon2 implementation|
|64|disable AVX2 Argon2 implementation|

**Pull request merged**